wallet: simplify and batch zap wallet txes process

furszy opened the pull request:
Work decoupled from #28574. Brother of #28894.

Includes two different, yet interconnected, performance and code improvements to the zap wallet transactions process.

1) As the goal of the `ZapSelectTx` function is to erase tx records that match any of the inputted hashes. There is no need to traverse the whole database record by record. We could just check if the tx exist, and remove it directly by calling `EraseTx()`.

2) Instead of performing single write operations per removed tx record, this PR batches them all within a single atomic db txn.

Moreover, these changes will enable us to consolidate all individual write operations that take place during the wallet migration process into a single db txn in the future.

furszy committed:
wallet: batch and simplify ZapSelectTx process

The goal of the function is to erase the wallet transactions that
match the inputted hashes. There is no need to traverse the database,
reading record by record, to then perform single entry removals for
each of them.

To ensure consistency and improve performance, this change-set removes
all tx records within a single atomic db batch operation, as well as
it cleans up code, improves error handling and simplifies the
transactions removal process entirely.

This optimizes the removal of watch-only transactions during the wallet
migration process and the 'removeprunedfunds' RPC command.

furszy committed:
wallet: ZapSelectTx, remove db rewrite code

The function does not return DBErrors::NEED_REWRITE.

furszy committed:
wallet: migration, remove extra NotifyTransactionChanged call

The wallet is unloaded at the beginning of the migration process,
so no object is listening to the signals.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2311,18 +2311,24 @@ DBErrors CWallet::LoadWallet()
 }
 
-DBErrors CWallet::ZapSelectTx(std::vector<uint256>& vHashIn)
+void CWallet::ClearTxns(const std::vector<uint256>& tx_hashes)
```
In 559529480fedd4181dab76f4452b5998d0cdadc3 "refactor: decouple wallet EraseTxns from ZapSelectTx"

The commit message calls this `EraseTxns`.

furszy replied:
fixed

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2311,18 +2311,24 @@ DBErrors CWallet::LoadWallet()
         const auto& it = mapWallet.find(hash);
         wtxOrdered.erase(it->second.m_it_wtxOrdered);
-        for (const auto& txin : it->second.tx->vin)
+        for (const auto& txin: it->second.tx->vin)
```
In 559529480fedd4181dab76f4452b5998d0cdadc3 "refactor: decouple wallet EraseTxns from ZapSelectTx"

Unnecessary whitespace change.

furszy replied:
fixed

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3931,10 +3933,26 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
     watchonly_batch.reset(); // Flush
     // Do the removes
     if (txids_to_delete.size() > 0) {
-        if (ZapSelectTx(txids_to_delete) != DBErrors::LOAD_OK) {
```
In ff214259de28216a5190c358b786306985f30227 "wallet: migration, remove watch-only transactions atomically"

It seems unnecessary to be making this change here as `ZapSelectTx` is now atomic.

furszy replied:
> In [ff21425](https://github.com/bitcoin/bitcoin/commit/ff214259de28216a5190c358b786306985f30227) "wallet: migration, remove watch-only transactions atomically"
> 
> It seems unnecessary to be making this change here as `ZapSelectTx` is now atomic.

True. This PR does not requires it. Will re-introduced it within #28574.

furszy reviewed (commented):
Updated per feedback. Thanks achow. [Diff](https://github.com/bitcoin/bitcoin/compare/ff214259de28216a5190c358b786306985f30227..d26a28508d5322fe7c29b3642dcf3a17b0ea31f1).

furszy reviewed (commented):
CI failure is unrelated. Relates to #29112.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2326,16 +2326,18 @@ void CWallet::ClearTxns(const std::vector<uint256>& tx_hashes)
+    if (!batch.TxnBegin()) return DBErrors::NONCRITICAL_ERROR;
+    DBErrors db_status = batch.ZapSelectTx(vHashIn);
+    // All deletions should have been added to the db txn, otherwise we abort the process.
+    if (db_status != DBErrors::LOAD_OK) return db_status;
```
In d26a28508d5322fe7c29b3642dcf3a17b0ea31f1 "wallet: batch ZapSelectTx db operations"

This should explicitly call `TxnAbort` before returning as otherwise we will get extraenous logs about the tx being aborted before batch is destroyed.

furszy replied:
> This should explicitly call TxnAbort before returning as otherwise we will get extraenous logs about the tx being aborted before batch is destroyed.

sure. Fixed.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2326,16 +2326,18 @@ void CWallet::ClearTxns(const std::vector<uint256>& tx_hashes)
+    if (db_status != DBErrors::LOAD_OK) return db_status;
 
+    // Apply db changes and remove transactions from the memory map
+    if (!batch.TxnCommit()) return DBErrors::NONCRITICAL_ERROR;
```
In d26a28508d5322fe7c29b3642dcf3a17b0ea31f1 "wallet: batch ZapSelectTx db operations"

I'm not sure if `NONCRITICAL_ERROR` is really the right return code. It seems like something has gone critically wrong if `TxnCommit` were to fail.

furszy replied:
> I'm not sure if `NONCRITICAL_ERROR` is really the right return code. It seems like something has gone critically wrong if `TxnCommit` were to fail.

The returned error code doesn't really matter. Anything different from `LOAD_OK` is an error for all the function's callers. And they behave exactly the same for all codes.
Actually, the `DBErrors` return doesn't make sense anymore after this PR changes. The function is no longer traversing the entire db nor checking for extra, unneeded, stuff like the version inside anymore.

Based on this, I have reworked the PR cleaning up this aspect as well.

furszy committed:
scripted-diff: rename ZapSelectTx to RemoveTxs

-BEGIN VERIFY SCRIPT-
sed -i 's/ZapSelectTx/RemoveTxs/g' $(git grep -l 'ZapSelectTx' ./src/wallet)
-END VERIFY SCRIPT-

furszy commented on src/wallet/walletdb.cpp:
```diff
@@ -1230,88 +1230,21 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
-            }
-            vTxHashOut.push_back(hash);
+        if (!EraseTx(hash)) {
+            LogPrint(BCLog::WALLETDB, "%s: Failure removing transaction: %s\n", __func__, hash.GetHex());
```
> Any reason to add __func__ to this log, when it previously wasn't there?

just an old bad habit. Removed. Thanks.

furszy reviewed (commented):
Updated per feedback. Thanks Marko and achow.

achow101 commented:
ACK 9a3c5c8697659e34d0476103af942a4615818f4e

The new code is much simpler.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/28987/commits/9a3c5c8697659e34d0476103af942a4615818f4e

looks good, also `RemoveTxs` is a much better name. Might be worth updating the description to mention the rename since the description ends up in the commit log.

hebasto commented on src/wallet/wallet.cpp:
```diff
@@ -3935,19 +3951,10 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
-        if (ZapSelectTx(txids_to_delete, deleted_txids) != DBErrors::LOAD_OK) {
-            error = _("Error: Could not delete watchonly transactions");
+        if (auto res = RemoveTxs(txids_to_delete); !res) {
+            error = _("Error: Could not delete watchonly transactions. ") + util::ErrorString(res);
```
nit: The trailing space could be easily overlooked and dropped during translation.

furszy replied:
> nit: The trailing space could be easily overlooked and dropped during translation.

@hebasto, what about creating a constant to represent the whitespace, or alternatively, implementing a function that concatenates strings with the whitespace inserted between them?
